   ##                                 CLOUD BASED INSURANCE DATABASE             

# ABSTRACT:

The idea behind this project is to build and Database that's almost simialar to the setup of an Insurance Service Provider's Enterprise Level Database/Data Repository. Since we are dealing with insurance data which holds sensitive personal details like address, age, sex, etc. obtaining a real world dataset is surreal, so, much of the data here are masked. We shall have also included an additional piece of code that shall be used to scrape the data from Twitter using Twitter API to allow our database to handle a real-world situation too. Here we are obtaining data from different sources and will be transforming them to fit into a uniform dataset template. This process is called as munging and post munging we will be cleaning the data to make sure that the data from those munged and cleaned dataset will be error free which can be loaded into a database.

The key take-away from this project is not just to munge the data and load into database, but also hold the entire database in a public cloud platfrom like Amazon Web Services Platform. For the pupose of this project we have built the database in AWS RDS, a AWS Database service and configured the connections and setup to which we shall load the data post munging it.

In [2]:
#importing all the packages we shall be using in this project to read the CSV file and munging the data and loading it into the MySQL Database
from __future__ import print_function
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#Tweepy packages are used here to allow us to pull the data from Twitter API
import tweepy
from tweepy.streaming import StreamListener
from tweepy import OAuthHandler
from tweepy import Stream
import json
from dateutil import parser
#MySQLdb package is used here to load the munged data into Database.
import MySQLdb
import re
from pandas.io import sql
import pprint

In [2]:
#obtaining the raw masked dataset.
ins = pd.read_csv('insurancedata.csv')
ins.head(5) # displaying top 5 rows

customerID  age age_category     sex     bmi  SeniorCitizen  children  \
0  7590-VHVEG   19        Minor  female  27.900              0         0   
1  5575-GNVDE   18        Minor    male  33.770              0         1   
2  3668-QPYBK   28        Major    male  33.000              0         3   
3  7795-CFOCW   33        Major    male  22.705              0         0   
4  9237-HQITU   32        Major    male  28.880              0         0   

  smoker  NUMBER              STREET  ...        Plan ID  policy_bind_date  \
0    yes    1111     Country Club Rd  ...  54192IN0020018        7/16/2002   
1     no      51           Parish Dr  ...  42261UT0070001       11/28/2002   
2     no      90  Stockings Brook Rd  ...  40572FL0070006        5/12/2007   
3     no      99      Lamentation Dr  ...  37903AR0070025        2/10/1998   
4     no     207      Lamentation Dr  ...  88380VA0720018         3/1/2012   

                                 Plan Marketing Name Plan Type  \
0  CareSource Marketplace Low Deductible Silver D...       HMO   
1  Healthy Premier Bronze w/3 Copays before Deduc...       EPO   
2                               Oscar Classic Silver       EPO   
3   Ambetter Balanced Care 7 (2020) (QualChoiceLife)       PPO   
4                 Anthem HealthKeepers Bronze X 5250       HMO   

  policy_bind_date.1 months_as_customer  umbrella_limit total_claim_amount  \
0          7/16/2002                  5             0.0              96200   
1         11/28/2002                462             0.0              31200   
2          5/12/2007                198             0.0              14500   
3          2/10/1998                384       1000000.0               7500   
4           3/1/2012                100       4000000.0              16500   

  report_available  claim_date  
0              YES    2/6/2015  
1                ?   1/18/2015  
2                ?   2/13/2015  
3               NO   1/27/2015  
4              YES   2/21/2015  

[5 rows x 35 columns]

We have a dataset, that has column values with special charecters like in the case of report_available, which needs to be handled and also check for duplicate values that hold no meaning and are redudant.

In [3]:
# we check the size of dataset that we are dealing with.
ins.shape

(299, 35)

The raw CSV that we are dealing here has about 35 columns and 299 Records

In [4]:
#For Audit purpose let's check if our dataset holds any null values.
ins.isna().sum()

customerID                     0
age                            0
age_category                   0
sex                            0
bmi                            0
SeniorCitizen                  0
children                       0
smoker                         0
NUMBER                         0
STREET                         0
state                          3
region                         0
self_employed                 11
insured_occupation             0
family_history                 0
PaymentMethod                  0
Enrollment charges             0
Average Monthly Tax Credit     0
Average Yearly Tax Credit      0
Medicaid Enrollment            0
Medicaid Enrollment ID        18
Medicare Enrollment ID         0
Metal Level                    0
Issuer Name                    0
HIOS Issuer ID                 0
Plan ID                        0
policy_bind_date               0
Plan Marketing Name            0
Plan Type                      0
policy_bind_date.1             0
months_as_

Before loading the data into Database, we need to handle those null fields values, in real-world scenario we can either choose to replace them or delete them depending upon the business case.

In [5]:
#To increase the quality of the data, we will be dropping the records having any null values
ins = ins.dropna()
ins.shape

(269, 35)

## Insurance Database

Now let us try to view the our Database Schema to understand, what are the tables we shall be building in to form our Insurance Database. We use the Entity-Relationship(ER) diagram to visualize our database

<img src= "Copy ofInsurance.jpeg"/>

## Creating DataFrames following the same schema of the corresponding tables

In [6]:
# Now we shall create dataframes
policy = pd.DataFrame(columns = ['Plan_ID','Medicaid_Enrollment','Medicaid_Enrollment_ID','Medicare_Enrollment_ID',
                                 'Plan_Marketing_Name','Plan_Type','Provider_ID'])
insu_provider = pd.DataFrame(columns = ['Provider_ID','Issuer_Name','HIOS_Issuer_ID','Metal_Level'])
claims = pd.DataFrame(columns = ['Claim ID','Plan_ID','Umbrella_Limit','Total_Claim_Amount','Report_Available'])
customer_address = pd.DataFrame(columns = ['Address_ID','Number','Street','State','Region','customerID'])
payment = pd.DataFrame(columns = ['Payment_ID','Plan_ID','Payment_Method','Enrollment_Charges',
                                  'Average_Monthly_Tax_Credit','Average_Yearly_Tax_Credit','customerID'])
#https://stackoverflow.com/questions/34682828/extracting-specific-selected-columns-to-new-dataframe-as-a-cop

# 1 NF Normalization:

Now we shall try to achive 1NF from on the Game Inventory by adding Primary Key. The same shall be done for all the tables to achieve the 1NF across the tables in the database that we are building.

# Customer Table:

Filtering the ins dataframe for customer details using filter function in pandas

In [7]:
customer =  pd.DataFrame(columns = ['customerID','age','sex','children','SeniorCitizen','family_history','months_as_customer','age_category','Plan ID'])
customer = ins.filter(['customerID','age','sex','children','SeniorCitizen','family_history','months_as_customer','age_category','Plan ID '],axis =1)
# Renaming the 'Plan ID ' column to 'Plan_ID'
customer = customer.rename(columns={'Plan ID ': 'Plan_ID'})
customer.head()

customerID  age     sex  children  SeniorCitizen family_history  \
12  8091-TTVAX   23    male         0              0            Yes   
13  0280-XJGEX   56  female         0              0             No   
14  5129-JLPIS   27    male         0              0             No   
15  3655-SNQYZ   19    male         1              0             No   
16  8191-XWSZG   52  female         1              0            Yes   

    months_as_customer age_category         Plan_ID  
12                 176        Major  60224GA0020005  
13                  79        Major  33653ME0050002  
14                 169        Major  29276OH0920418  
15                  89        Minor  20129IL0330020  
16                  85        Major  37833WI0380028

## Policy Table:

In [8]:
policy['Plan_ID'] = ins['Plan ID '].unique()
# since the dataset that we handle here has already a column Plan_ID which we shall make as our Primary Key, by ensuring that each record in that column is unique
med_enro = []
med_enro_id = []
med_care_id = []
plan_mark_name = []
plan_type = []
# Now that we have the table schema defined, we shall pull the records from our ins Dataframe into appropriate columns
for i in policy['Plan_ID']:
    value = ins.loc[(ins['Plan ID '] == i),'Medicaid Enrollment'].tolist()[0]
    value1 = ins.loc[(ins['Plan ID '] == i),'Medicaid Enrollment ID'].tolist()[0]
    value2 = ins.loc[(ins['Plan ID '] == i),'Medicare Enrollment ID'].tolist()[0]
    value3 = ins.loc[(ins['Plan ID '] == i),'Plan Marketing Name'].tolist()[0]
    value4 = ins.loc[(ins['Plan ID '] == i),'Plan Type'].tolist()[0]
    med_enro.append(value)
    med_enro_id.append(value1)
    med_care_id.append(value2)
    plan_mark_name.append(value3)
    plan_type.append(value4)
policy['Medicaid_Enrollment'] = med_enro
policy['Medicaid_Enrollment_ID'] = med_enro_id
policy['Medicare_Enrollment_ID'] = med_care_id 
policy['Plan_Marketing_Name'] = plan_mark_name
policy['Plan_Type'] = plan_type
# Displaying the top 5 records of Policy table
policy.head()

Plan_ID  Medicaid_Enrollment  Medicaid_Enrollment_ID  \
0  60224GA0020005                False               9935517.0   
1  33653ME0050002                 True               5727510.0   
2  29276OH0920418                 True               1220788.0   
3  20129IL0330020                 True               2748165.0   
4  37833WI0380028                False               4123280.0   

   Medicare_Enrollment_ID                                Plan_Marketing_Name  \
0                  294284  CareSource Marketplace Low Deductible Silver D...   
1                 2118300                              Community Advance PPO   
2                 1181014             Anthem Silver Pathway X HMO 10 for HSA   
3                  587780                        HMO 8150 Elite Catastrophic   
4                  499753                 Quartz One Silver I302 with Dental   

  Plan_Type Provider_ID  
0       HMO         NaN  
1       PPO         NaN  
2       HMO         NaN  
3       HMO         NaN  
4       HMO         NaN

Note: That for the policy table we have column Provider_ID which has value NA loaded now. This is done to have an entity relationship between the Policy Table and the Provider Table. The kind of relation between the two entites is maintained as defined in the ER Diagram shown above.

## Provider Table:

Creating Provider table schema and loading the table.

In [9]:
insu_provider['Issuer_Name'] = ins['Issuer Name'].unique()
# here also we have a column Prov_Id from our raw csv, which we shall consider as primary key to our Provider Table, to ensure that the column loads with unique values we apply the unique function to the list
hios_temp = []
metal_temp = []
prov_id = []
j = 201

# Now we shall load the data from our ins dataframe into the appropriate columns
for i in insu_provider['Issuer_Name']:
    value = ins.loc[(ins['Issuer Name'] == i),'HIOS Issuer ID'].tolist()[0]
    value1 = ins.loc[(ins['Issuer Name'] == i),'Metal Level'].tolist()[0]
    hios_temp.append(value)
    metal_temp.append(value1)
    prov_id.append(j)
    j += 1
insu_provider['HIOS_Issuer_ID'] = hios_temp
insu_provider['Metal_Level'] = metal_temp
insu_provider['Provider_ID'] = prov_id
#displaying top 5 records
insu_provider.head()

Provider_ID                          Issuer_Name  HIOS_Issuer_ID  \
0          201               CareSource Georgia Co.           60224   
1          202             Community Health Options           33653   
2          203    Anthem Blue Cross and Blue Shield           29276   
3          204  Health Alliance Medical Plans, Inc.           20129   
4          205                               Quartz           37833   

    Metal_Level  
0        Silver  
1        Silver  
2        Silver  
3  Catastrophic  
4        Silver

## Claim Table:

Creating Claim table schema and loading the table.

In [10]:
umb_temp = []
j = 601# here to generate a ClaimId we shall consider the generating some random number starting from 601
tot_claim_temp = []
rep_avail = []
claim_id = []
claims['Plan_ID'] = ins['Plan ID '].unique()

#once the schema for the Claim table is ready we shall try to load the ins data into the claim table
for i in claims['Plan_ID']:
    value1 = ins.loc[(ins['Plan ID '] == i),'umbrella_limit'].tolist()[0]
    value2 = ins.loc[(ins['Plan ID '] == i),'total_claim_amount'].tolist()[0]
    value3 = ins.loc[(ins['Plan ID '] == i),'report_available'].tolist()[0]
    umb_temp.append(value1)
    tot_claim_temp.append(value2)
    rep_avail.append(value3)
    claim_id.append(j)
    j += 1
claims['Claim_ID'] = claim_id
claims['Umbrella_Limit'] = umb_temp
claims['Total_Claim_Amount'] = tot_claim_temp
claims['Report_Available'] = rep_avail
#displaying top 5 rows
claims.head()

Claim ID         Plan_ID  Umbrella_Limit  Total_Claim_Amount  \
0      NaN  60224GA0020005       1000000.0               11310   
1      NaN  33653ME0050002       2000000.0              108710   
2      NaN  29276OH0920418       2000000.0                2690   
3      NaN  20129IL0330020             0.0                1500   
4      NaN  37833WI0380028       1000000.0               71240   

  Report_Available  Claim-ID  
0                ?       601  
1                ?       602  
2              YES       603  
3               NO       604  
4                ?       605

In [11]:
customer.head()

customerID  age     sex  children  SeniorCitizen family_history  \
12  8091-TTVAX   23    male         0              0            Yes   
13  0280-XJGEX   56  female         0              0             No   
14  5129-JLPIS   27    male         0              0             No   
15  3655-SNQYZ   19    male         1              0             No   
16  8191-XWSZG   52  female         1              0            Yes   

    months_as_customer age_category         Plan_ID  
12                 176        Major  60224GA0020005  
13                  79        Major  33653ME0050002  
14                 169        Major  29276OH0920418  
15                  89        Minor  20129IL0330020  
16                  85        Major  37833WI0380028

## Customer_Address Table:

Creating Customer Address table schema and loading the table. 

In [12]:
customer_address['customerID'] = ins.customerID
customer_address['Number'] = ins['NUMBER']
customer_address['Street'] = ins.STREET
customer_address['State'] = ins.state
customer_address['Region'] = ins.region
j = 901
address_id = []
for i in customer_address['Number']:
    address_id.append(j)
    j += 1
customer_address.Address_ID = address_id
#displaying the top 5 rows
customer_address.head()

Address_ID  Number           Street State     Region  customerID
12         901      26      Magnolia Ln    NY  southwest  8091-TTVAX
13         902      61     Highview Ter    NC  southeast  0280-XJGEX
14         903     878  Chamberlain Hwy    MA  southeast  5129-JLPIS
15         904     782    Kensington Rd    IA  southwest  3655-SNQYZ
16         905      34      Magnolia Ln    CA  northeast  8191-XWSZG

## Payment Table:

Creating Payment table schema and loading the table.

In [13]:
payment['Plan_ID'] = ins['Plan ID '].unique()
pay_method = []
enro_charges = []
av_month_tax = []
av_year_tax = []
cust_id = []
payment_id = []
j = 301#here again we try to generate a random number for the PaymentId column starting from 301
for i in payment['Plan_ID']:
    value1 = ins.loc[(ins['Plan ID ']==i),'PaymentMethod'].tolist()[0]
    value2 = ins.loc[(ins['Plan ID ']==i),'Enrollment charges'].tolist()[0]
    value3 = ins.loc[(ins['Plan ID ']==i), 'Average Monthly Tax Credit'].tolist()[0]
    value4 = ins.loc[(ins['Plan ID ']==i), 'Average Yearly Tax Credit'].tolist()[0]
    value5 = ins.loc[(ins['Plan ID ']==i), 'customerID'].tolist()[0]
    pay_method.append(value1)
    enro_charges.append(value2)
    av_month_tax.append(value3)
    av_year_tax.append(value4)
    cust_id.append(value5)
    payment_id.append(j)
    j += 1
# we pull each field value of a row and append intp the corresponding list 
payment['Payment_Method'] = pay_method
payment['Enrollment_Charges'] = enro_charges
payment['Average_Monthly_Tax_Credit'] = av_month_tax
payment['Average_Yearly_Tax_Credit'] = av_year_tax
payment['customerID'] = cust_id
payment['Payment_ID'] = payment_id
#displaying the top 5 rows
payment.head()
payment.tail()

Payment_ID        Plan_ID    Payment_Method  Enrollment_Charges  \
253         554  392C445WQ5267      Mailed check           3906.1270   
254         555  701F130FY9197      Mailed check           1704.5681   
255         556  748D434TK3215  Electronic check          16297.8460   
256         557  864Y799RR6253  Electronic check          21978.6769   
257         558  100W188OU7332  Electronic check          38746.3551   

    Average_Monthly_Tax_Credit Average_Yearly_Tax_Credit  customerID  
253                       $293                     $3516  1563-IWQEX  
254                       $146                     $1752  8203-XJZRC  
255                       $377                     $4524  6556-DBKZF  
256                       $215                     $2580  6851-WEFYX  
257                       $236                     $2832  2985-JUUBZ

# 2 Normalization:

In our Payments table we have a column Average_Yearly_Tax_Credit, which is calculated from the Average_Monthly_Tax_Credit * 12
To achieve 2NF form, we should not have any calcuated columns. Hence, we remove the Average_Yearly_Tax_Credit from the table.

In [14]:
payment = payment.drop(columns=['Average_Yearly_Tax_Credit'])
payment.head()

Payment_ID         Plan_ID             Payment_Method  Enrollment_Charges  \
0         301  60224GA0020005    Credit card (automatic)             1826.84   
1         302  33653ME0050002  Bank transfer (automatic)            11090.72   
2         303  29276OH0920418           Electronic check            39611.76   
3         304  20129IL0330020    Credit card (automatic)             1837.24   
4         305  37833WI0380028               Mailed check            10797.34   

  Average_Monthly_Tax_Credit  customerID  
0                      $265   8091-TTVAX  
1                      $237   0280-XJGEX  
2                      $259   5129-JLPIS  
3                      $307   3655-SNQYZ  
4                      $247   8191-XWSZG

In [15]:
customer.tail(10)

customerID  age     sex  children  SeniorCitizen family_history  \
289  3580-REOAC   52    male         3              0             No   
290  7534-BFESC   28  female         0              1             No   
291  3727-OWVYD   29    male         1              0             No   
292  2294-SALNE   25    male         2              0            Yes   
293  4847-TAJYI   22  female         0              1            Yes   
294  1563-IWQEX   25    male         3              0             No   
295  8203-XJZRC   18    male         0              0             No   
296  6556-DBKZF   19    male         0              0            Yes   
297  6851-WEFYX   47    male         1              1            Yes   
298  2985-JUUBZ   31    male         3              0            Yes   

     months_as_customer age_category        Plan_ID  
289                 414        Major  160P883IN8477  
290                 316        Major  710U610AC4760  
291                 105        Major  969N739XY3979  
292                  93        Major  205K482FE6220  
293                 103        Major  364S829VL9586  
294                 141        Major  392C445WQ5267  
295                 256        Minor  701F130FY9197  
296                 177        Minor  748D434TK3215  
297                 245        Major  864Y799RR6253  
298                 186        Major  100W188OU7332

In our Customer Table we have a column age_category, which assigns if each customer is Minor or Major. This field is calculated based upon the age value, if a customer is less than 19, he/she is considered as minor else assigned value major. To achieve 2NF we shouldn't have any calculated field, since age_category is a calculated column we choose to drop the field.

In [16]:
customer = customer.drop(columns=['age_category'])
customer.head()

customerID  age     sex  children  SeniorCitizen family_history  \
12  8091-TTVAX   23    male         0              0            Yes   
13  0280-XJGEX   56  female         0              0             No   
14  5129-JLPIS   27    male         0              0             No   
15  3655-SNQYZ   19    male         1              0             No   
16  8191-XWSZG   52  female         1              0            Yes   

    months_as_customer         Plan_ID  
12                 176  60224GA0020005  
13                  79  33653ME0050002  
14                 169  29276OH0920418  
15                  89  20129IL0330020  
16                  85  37833WI0380028

In [ ]:
policy.tail(10)

# 3 Normalization:


To achieve the 3NF each column in the table should have direct relation to the PK. We have ID columns of other tables which are kept to establish a relationship to another table, this is called the Primarykey-Foreignkey relation. Here in our project, we show such relationship using the Policy and Provider table. An Insurance provider shall have many policies under its plan, so each Policy in the Policy table will be linked to an Insurance provider in the Provider table. 
This is what achieved from the code below.

In [17]:
### Adding the generated provider id to corresponding columns in policy
prov_id_temp =[]
#here  we first pull the Provider Ids from the Provider table and try to load into the empty list
for i in ins['Issuer Name']:
    values = insu_provider.loc[insu_provider['Issuer_Name'] == i,'Provider_ID'].tolist()[0]
    prov_id_temp.append(values)
ins['Provider ID'] = prov_id_temp
# now the pulled provider id details are assigned to each rows of the Policy Table to achieve the 3NF 
prov_id_new = []
for i in policy['Plan_ID']:
    values = ins.loc[(ins['Plan ID '] == i),'Provider ID'].tolist()[0]
    prov_id_new.append(values)
policy['Provider_ID'] = prov_id_new
#
policy.tail(10)

Plan_ID  Medicaid_Enrollment  Medicaid_Enrollment_ID  \
248  160P883IN8477                 True                189429.0   
249  710U610AC4760                 True                918868.0   
250  969N739XY3979                False                788544.0   
251  205K482FE6220                 True               5146113.0   
252  364S829VL9586                 True               6357489.0   
253  392C445WQ5267                 True               9239334.0   
254  701F130FY9197                False                180874.0   
255  748D434TK3215                False               1766047.0   
256  864Y799RR6253                False                693869.0   
257  100W188OU7332                False               2072282.0   

     Medicare_Enrollment_ID  \
248                 7235717   
249                 2561709   
250                 8130629   
251                 4009357   
252                 4381867   
253                 5357375   
254                 1873043   
255                 6636223   
256                  793561   
257                 5401092   

                                   Plan_Marketing_Name Plan_Type  Provider_ID  
248                 my Blue Access EPO Silver 3950 HSA       EPO          220  
249                          BlueOptions Platinum 1424       EPO          251  
250                   Anthem Bronze Pathway X HMO 5200       HMO          213  
251  my Blue Access WV Major Events EPO 8150 - 3 Fr...       EPO          254  
252  my Blue Access WV EPO Gold 800 - 2 Free PCP Vi...       EPO          213  
253  CareSource Marketplace Gold Dental, Vision, & ...       HMO          220  
254  Ambetter Secure Care 5 (2020) + Vision + Adult...       EPO          240  
255  my Blue Access WV EPO Gold 800 - 2 Free PCP Vi...       EPO          265  
256                               IND Bronze HMO  7500       HMO          206  
257  CareSource Marketplace Low Deductible Silver D...       HMO          203

# Social Media Scrapping

To use real-time data in our project, we have scrapped the Twitter's Tweets using the Twitter API provided by the service provider.

In [18]:
consumer_key = ''
consumer_secret = ''
access_token = ''
access_token_secret = ''
#these keys are all shared by the twitter api to scrap the tweets.

# Connecting to AWS RDS Database

Since the primary goal of the project it to have a database setup in virtual cloud platform, we now connect to the AWS RDS database that holds our Insurance Database.

In [1]:
HOST = 'dmdd.cep3r0xw222loxj.us-east-21.rds.amazonaws.com'
USER = 'admini'
PASSWD = 'superman******12345'
DATABASE = 'aws_insurance'
# all the connection details have been masked for security purpose.

Connecting to Twitter to scrap the Tweets and Twitter User data

In [20]:
#we now shall try to pull the data for our social media, by scrapping in tweets from Twitter 
def store_data(created_at, text, screen_name, tweet_id,location,followers_count):
    db=MySQLdb.connect(host=HOST, user=USER, passwd=PASSWD, db=DATABASE, charset="utf8")
    cursor = db.cursor()
    insert_query = "INSERT INTO twitter (tweet_id, screen_name,created_at, text, location,follower_count) VALUES (%s, %s, %s, %s, %s, %s)"
    cursor.execute(insert_query, (tweet_id, screen_name, created_at, text, location,followers_count))
    db.commit()
    cursor.close()
    db.close()
    return
class StreamListener(tweepy.StreamListener):    
    #This is a class provided by tweepy to access the Twitter Streaming API. 

    def on_connect(self):
        # Called initially to connect to the Streaming API
        print("You are now connected to the streaming API.")
 
    def on_error(self, status_code):
        # On error - if an error occurs, display the error / status code
        print('An Error has occured: ' + repr(status_code))
        return False
 
    def on_data(self, data):
        #This is the meat of the script...it connects to your database and stores the tweet
        try:
           # Decode the JSON from Twitter
            datajson = json.loads(data)
            
            #grab the wanted data from the Tweet
            text = datajson['text']
            screen_name = datajson['user']['screen_name']
            tweet_id = datajson['id']
            created_at = parser.parse(datajson['created_at'])
            #hashtag = re.findall(r"#(\w+)",datajson['text'])
            location = datajson['user']['location']
            followers_count = datajson['user']['followers_count']

            #print out a message to the screen that we have collected a tweet
            print("Tweet collected at " + str(created_at))
            
            #insert the data into the MySQL database
            store_data(created_at, text, screen_name, tweet_id,location,followers_count)
        
        except Exception as e:
           print(e)

Now we shall try creating a schema for the twitter table that we scrapped

In [21]:
db = MySQLdb.connect(host=HOST, user=USER, passwd=PASSWD, db=DATABASE, charset="utf8")
print (db)
cursor = db.cursor()
#Drop the twitter table
#cursor.execute("drop table `twitter`")

# create a table called twitter
cursor.execute("""CREATE TABLE `twitter` (
`id` int(11) NOT NULL AUTO_INCREMENT,
 `tweet_id` varchar(250) DEFAULT NULL,
 `screen_name` varchar(128) DEFAULT NULL,
 `created_at` timestamp NULL DEFAULT NULL,
 `text` text,
 `location` varchar(128) DEFAULT NULL,
 `follower_count` int(128) DEFAULT NULL,
 PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;""")

<_mysql.connection open to 'dmdd.cep3r0xwloxj.us-east-2.rds.amazonaws.com' at 0000005F319DC988>


0

## Scrapping Tweets that contains keywords:

Here we scrape the tweet data from twitter that contains hastag like '#medicaid','#medicare','#usmedicalinsurance','#usinsurance' etc
for our Tweets table

In [23]:
consumer_key = 'e6a6AL8uzdrq0rk8gAUpkf9Am'
consumer_secret = 'lQ6qhk64Ai8rdjC2Ba5lMLAaUk5dfZQog0gVwUXkcgk9j5e5eU'
access_token = '367199610-hYrg1EKMN8k3gzKibErKzQMzIP8viYS1l204yPKc'
access_token_secret = 'HAo8dEOfosISGRiExd07ILr8KToL3zphpXfqomLW5e1ew'
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
#Set up the listener. The 'wait_on_rate_limit=True' is needed to help with Twitter API rate limiting.
listener = StreamListener(api=tweepy.API(wait_on_rate_limit=True)) 
streamer = tweepy.Stream(auth=auth, listener=listener)
tags = ['#medicaid','#medicare','#usmedicalinsurance','#usinsurance','medic aid','medicare','medicaid','medic care']
print("Tracking: " + str(tags))
streamer.filter(track=tags)

Tracking: ['#medicaid', '#medicare', '#usmedicalinsurance', '#usinsurance', 'medic aid', 'medicare', 'medicaid', 'medic care']
You are now connected to the streaming API.
Tweet collected at 2020-04-24 22:13:40+00:00
Tweet collected at 2020-04-24 22:13:41+00:00
Tweet collected at 2020-04-24 22:13:48+00:00
Tweet collected at 2020-04-24 22:14:09+00:00
Tweet collected at 2020-04-24 22:14:10+00:00
Tweet collected at 2020-04-24 22:14:15+00:00
Tweet collected at 2020-04-24 22:14:19+00:00
Tweet collected at 2020-04-24 22:14:20+00:00
Tweet collected at 2020-04-24 22:14:21+00:00
Tweet collected at 2020-04-24 22:14:27+00:00
Tweet collected at 2020-04-24 22:14:32+00:00
Tweet collected at 2020-04-24 22:14:41+00:00
Tweet collected at 2020-04-24 22:14:41+00:00
Tweet collected at 2020-04-24 22:14:46+00:00
Tweet collected at 2020-04-24 22:14:52+00:00
Tweet collected at 2020-04-24 22:15:03+00:00
Tweet collected at 2020-04-24 22:15:06+00:00
Tweet collected at 2020-04-24 22:15:07+00:00
Tweet collected at 

Tweet collected at 2020-04-24 22:24:39+00:00
Tweet collected at 2020-04-24 22:24:46+00:00
Tweet collected at 2020-04-24 22:24:56+00:00
Tweet collected at 2020-04-24 22:24:58+00:00
Tweet collected at 2020-04-24 22:25:00+00:00
Tweet collected at 2020-04-24 22:25:03+00:00
Tweet collected at 2020-04-24 22:25:06+00:00
Tweet collected at 2020-04-24 22:25:06+00:00
Tweet collected at 2020-04-24 22:25:06+00:00
Tweet collected at 2020-04-24 22:25:09+00:00
Tweet collected at 2020-04-24 22:25:16+00:00
Tweet collected at 2020-04-24 22:25:16+00:00
Tweet collected at 2020-04-24 22:25:17+00:00
Tweet collected at 2020-04-24 22:25:27+00:00
Tweet collected at 2020-04-24 22:25:28+00:00
Tweet collected at 2020-04-24 22:25:32+00:00
Tweet collected at 2020-04-24 22:25:35+00:00
Tweet collected at 2020-04-24 22:25:43+00:00
Tweet collected at 2020-04-24 22:25:45+00:00
Tweet collected at 2020-04-24 22:25:45+00:00
Tweet collected at 2020-04-24 22:25:51+00:00
Tweet collected at 2020-04-24 22:25:55+00:00
Tweet coll

Tweet collected at 2020-04-24 22:35:39+00:00
Tweet collected at 2020-04-24 22:35:50+00:00
Tweet collected at 2020-04-24 22:35:50+00:00
Tweet collected at 2020-04-24 22:35:52+00:00
Tweet collected at 2020-04-24 22:35:58+00:00
Tweet collected at 2020-04-24 22:35:59+00:00
Tweet collected at 2020-04-24 22:36:12+00:00
Tweet collected at 2020-04-24 22:36:16+00:00
Tweet collected at 2020-04-24 22:36:17+00:00
Tweet collected at 2020-04-24 22:36:22+00:00
Tweet collected at 2020-04-24 22:36:23+00:00
Tweet collected at 2020-04-24 22:36:28+00:00
Tweet collected at 2020-04-24 22:36:28+00:00
Tweet collected at 2020-04-24 22:36:32+00:00
Tweet collected at 2020-04-24 22:36:34+00:00
Tweet collected at 2020-04-24 22:36:34+00:00
Tweet collected at 2020-04-24 22:36:36+00:00
Tweet collected at 2020-04-24 22:36:37+00:00
Tweet collected at 2020-04-24 22:36:42+00:00
Tweet collected at 2020-04-24 22:36:44+00:00
Tweet collected at 2020-04-24 22:36:46+00:00
Tweet collected at 2020-04-24 22:36:55+00:00
Tweet coll

Tweet collected at 2020-04-24 22:48:04+00:00
Tweet collected at 2020-04-24 22:48:08+00:00
Tweet collected at 2020-04-24 22:48:17+00:00
Tweet collected at 2020-04-24 22:48:20+00:00
Tweet collected at 2020-04-24 22:48:24+00:00
Tweet collected at 2020-04-24 22:48:25+00:00
Tweet collected at 2020-04-24 22:48:28+00:00
Tweet collected at 2020-04-24 22:48:35+00:00
Tweet collected at 2020-04-24 22:48:36+00:00
Tweet collected at 2020-04-24 22:48:38+00:00
Tweet collected at 2020-04-24 22:48:38+00:00
Tweet collected at 2020-04-24 22:48:44+00:00
Tweet collected at 2020-04-24 22:48:49+00:00
Tweet collected at 2020-04-24 22:48:49+00:00
Tweet collected at 2020-04-24 22:48:53+00:00
Tweet collected at 2020-04-24 22:48:53+00:00
Tweet collected at 2020-04-24 22:48:55+00:00
Tweet collected at 2020-04-24 22:48:55+00:00
Tweet collected at 2020-04-24 22:49:00+00:00
Tweet collected at 2020-04-24 22:49:03+00:00
Tweet collected at 2020-04-24 22:49:10+00:00
Tweet collected at 2020-04-24 22:49:10+00:00
Tweet coll

Tweet collected at 2020-04-24 22:59:12+00:00
Tweet collected at 2020-04-24 22:59:15+00:00
Tweet collected at 2020-04-24 22:59:16+00:00
Tweet collected at 2020-04-24 22:59:28+00:00
Tweet collected at 2020-04-24 22:59:29+00:00
Tweet collected at 2020-04-24 22:59:32+00:00
Tweet collected at 2020-04-24 22:59:33+00:00
Tweet collected at 2020-04-24 22:59:46+00:00
Tweet collected at 2020-04-24 22:59:51+00:00
Tweet collected at 2020-04-24 22:59:57+00:00
Tweet collected at 2020-04-24 22:59:57+00:00
Tweet collected at 2020-04-24 23:00:01+00:00
Tweet collected at 2020-04-24 23:00:04+00:00
Tweet collected at 2020-04-24 23:00:07+00:00
Tweet collected at 2020-04-24 23:00:10+00:00
Tweet collected at 2020-04-24 23:00:12+00:00
Tweet collected at 2020-04-24 23:00:12+00:00
Tweet collected at 2020-04-24 23:00:20+00:00
Tweet collected at 2020-04-24 23:00:21+00:00
Tweet collected at 2020-04-24 23:00:31+00:00
Tweet collected at 2020-04-24 23:00:33+00:00
Tweet collected at 2020-04-24 23:00:36+00:00
Tweet coll

KeyboardInterrupt: 

# Creating Tables in AWS Cloud  from Python & Loading Data

We move to next step of creating the tables in our AWS Cloud setup.

In [22]:
from sqlalchemy import create_engine
#the sqlalchemy package is used to connect python to rds and create tables in the aws rds
engine = create_engine('mysql://admini:dmdd12345@dmdd.cep3r0xwloxj.us-east-21.rds.amazonaws.com/aws_insurance',encoding="utf-8")
#loading the Payment Table
sql.to_sql(payment, con=engine, name='payment', 
                if_exists='replace',index = False)
#loading the Claims Table
sql.to_sql(claims, con=engine, name='claims', 
                if_exists='replace',index = False)
#loading the Insurance Providers Table
sql.to_sql(insu_provider, con=engine, name='insuranceprovider', 
                if_exists='replace',index = False)
#loading the Policy Table
sql.to_sql(policy, con=engine, name='policy', 
                if_exists='replace',index = False)
#loading the Customer Table
sql.to_sql(customer, con=engine, name='customer', 
                if_exists='replace',index = False)
#loading the Customer Address Table
sql.to_sql(customer_address, con=engine, name='customer_address', 
                if_exists='replace',index = False)
ins_new = ins.rename(columns={'Plan ID ': 'Plan ID'})
#loading the RAW CSV into Table
sql.to_sql(ins_new, con=engine, name='insurance', 
                if_exists='replace',index = False)

We will be collecting twitter data of Issuers. We will be storing the names of the Issuer in an array object so that we can pass it to the twitter API to get the corresponding details of the issuer such as tweets, location, description of the company, tweets, number of followers, etc.

In [24]:
#checking for the Hastags if they contains any Insurance Provider related tags
issuer_names = insu_provider['Issuer_Name'].tolist()
print ('There are {} insurance providers. We will be collecting more information of the insurance providers using the Twitter API and storing it in the MySQL '.format(len(issuer_names)))

There are 69 insurance providers. We will be collecting more information of the insurance providers using the Twitter API and storing it in the MySQL 


In [25]:
#connecting to Twitter
consumer_key = 'e6a6AL8uzdrq0rk8gAUpkf9Am'
consumer_secret = 'lQ6qhk64Ai8rdjC2Ba5lMLAaUk5dfZQog0gVwUXkcgk9j5e5eU'
access_token = '367199610-hYrg1EKMN8k3gzKibErKzQMzIP8viYS1l204yPKc'
access_token_secret = 'HAo8dEOfosISGRiExd07ILr8KToL3zphpXfqomLW5e1ew'
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
#Set up the listener. The 'wait_on_rate_limit=True' is needed to help with Twitter API rate limiting.
listener = StreamListener(api=tweepy.API(wait_on_rate_limit=True)) 
streamer = tweepy.Stream(auth=auth, listener=listener)

Gathering the Twitter User table

In [26]:
# here we use the tweepy package to scrape the twitter user data directly from twitter
verified = []
twitter_id = []
tname = []
user_name = []
location = []
url = []
followers_count = []
statuses_count = []
provider_id = []
api = tweepy.API(auth)
for name in issuer_names:
    provider = insu_provider.loc[(insu_provider['Issuer_Name'] == name),'Provider_ID'].tolist()[0]
    user = api.search_users(name)
    if len(user) > 0 :
        provider_id.append(provider)
        verified.append(user[0].verified)
        twitter_id.append(user[0].id)
        tname.append(user[0].name)
        user_name.append(user[0].screen_name)
        location.append(user[0].location)
        url.append(user[0].url)
        followers_count.append(user[0].followers_count)
        statuses_count.append(user[0].statuses_count)

Creating Schema for the Twitter User Table:

In [27]:
tw_user = pd.DataFrame(columns = ['Twitter_ID', 'Name','Username', 'Verified',
                                  'Location','url','No_of_Followers','No_of_Tweets','Provider_ID'])
#assigning data to the schema
tw_user['Twitter_ID'] = twitter_id
tw_user['Name'] = tname
tw_user['Username'] = user_name
tw_user['Verified'] = verified
tw_user['Location'] = location
tw_user['url'] = url
tw_user['No_of_Followers'] = followers_count
tw_user['No_of_Tweets'] = statuses_count
tw_user['Provider_ID'] = provider_id

## Twitter User Table:

Loading the scrapped user into Twitter table

In [28]:
sql.to_sql(tw_user, con=engine, name='provider_twitter', 
                if_exists='replace',index = False)

In [29]:
tw_user.head()

Twitter_ID                  Name         Username  Verified  \
0   849775556        Health Options    CmtyHlthOptns     False   
1   467554336      Anthem BCBS News  AnthemBCBS_News     False   
2   573918122                Quartz               qz      True   
3    27922157                   AMA  AmerMedicalAssn      True   
4    20519602  FirstCareHealthPlans        firstcare     False   

                  Location                      url  No_of_Followers  \
0          Lewiston, Maine   http://t.co/PPlbfyzSyj              602   
1                            http://t.co/8eY95I0HOl             3372   
2                The World  https://t.co/gLfIrRbtbb           388622   
3  Chicago/Washington D.C.   http://t.co/MEg0g5QqL2           722458   
4                    Texas   http://t.co/jK5HXsqTnu              298   

   No_of_Tweets  Provider_ID  
0          3741          202  
1          9415          203  
2        176897          205  
3         27748          206  
4          1022          209

Scrapping the Tweets data

In [30]:
import twitter
#we use the twitter package again to scrap the twitter user tweets  data
api = tweepy.API(auth)
tw_id = []
twt_id = []
twt_text = []
no_retweets = []
fav_count = []
j = 0
#capturing the user data in real-time
for scrname in user_name:
    try:
        tweets = api.user_timeline(screen_name=scrname, count=10)
    except:
        pass
    for tweet in tweets:
        tw_id.append(tweet.user.id)
        twt_id.append(tweet.id)
        twt_text.append(tweet.text)
        no_retweets.append(tweet.retweet_count)
        fav_count.append(tweet.favorite_count)
    j +=1

## Tweets Table:

Creating Schema for Tweets Table:

In [31]:
tweets_ = pd.DataFrame(columns = ['Tweet_ID','Twitter_ID', 'Text','No_of Retweets', 'Favourites_Count'])
tweets_['Tweet_ID'] = twt_id
tweets_['Twitter_ID'] = tw_id
tweets_['Text'] = twt_text
tweets_['No_of_Retweets'] = no_retweets
tweets_['Favourites_Count'] = fav_count

In [32]:
#connection_string = f"admini:dmdd1234@dmdd.cep3r0xwloxj.us-east-2.rds.amazonaws.com:3306/proj_insurance?charset=utf8"
#engine = create_engine(f'mysql://{connection_string}')
engine = create_engine('mysql://admini:dmdd12345@dmdd.cep3r0xwloxj.us-east-21.rds.amazonaws.com/aws_insurance?charset=utf8')
#for security purpose we have masked the url connection to aws

Loading the Tweets table

In [33]:
sql.to_sql(tweets_, con=engine, name='tweets', 
                if_exists='replace',index = False)

Final Structure of all the Tables for our Insurance Database

In [34]:
insu_provider.head(5)

Provider_ID                          Issuer_Name  HIOS_Issuer_ID  \
0          201               CareSource Georgia Co.           60224   
1          202             Community Health Options           33653   
2          203    Anthem Blue Cross and Blue Shield           29276   
3          204  Health Alliance Medical Plans, Inc.           20129   
4          205                               Quartz           37833   

    Metal_Level  
0        Silver  
1        Silver  
2        Silver  
3  Catastrophic  
4        Silver

In [35]:
policy.head(5)

Plan_ID  Medicaid_Enrollment  Medicaid_Enrollment_ID  \
0  60224GA0020005                False               9935517.0   
1  33653ME0050002                 True               5727510.0   
2  29276OH0920418                 True               1220788.0   
3  20129IL0330020                 True               2748165.0   
4  37833WI0380028                False               4123280.0   

   Medicare_Enrollment_ID                                Plan_Marketing_Name  \
0                  294284  CareSource Marketplace Low Deductible Silver D...   
1                 2118300                              Community Advance PPO   
2                 1181014             Anthem Silver Pathway X HMO 10 for HSA   
3                  587780                        HMO 8150 Elite Catastrophic   
4                  499753                 Quartz One Silver I302 with Dental   

  Plan_Type  Provider_ID  
0       HMO          201  
1       PPO          202  
2       HMO          203  
3       HMO          204  
4       HMO          205

In [36]:
claims.head(5)

Claim ID         Plan_ID  Umbrella_Limit  Total_Claim_Amount  \
0      NaN  60224GA0020005       1000000.0               11310   
1      NaN  33653ME0050002       2000000.0              108710   
2      NaN  29276OH0920418       2000000.0                2690   
3      NaN  20129IL0330020             0.0                1500   
4      NaN  37833WI0380028       1000000.0               71240   

  Report_Available  Claim-ID  
0                ?       601  
1                ?       602  
2              YES       603  
3               NO       604  
4                ?       605

In [37]:
customer.head(5)

customerID  age     sex  children  SeniorCitizen family_history  \
12  8091-TTVAX   23    male         0              0            Yes   
13  0280-XJGEX   56  female         0              0             No   
14  5129-JLPIS   27    male         0              0             No   
15  3655-SNQYZ   19    male         1              0             No   
16  8191-XWSZG   52  female         1              0            Yes   

    months_as_customer         Plan_ID  
12                 176  60224GA0020005  
13                  79  33653ME0050002  
14                 169  29276OH0920418  
15                  89  20129IL0330020  
16                  85  37833WI0380028

In [38]:
customer_address.head(5)

Address_ID  Number           Street State     Region  customerID
12         901      26      Magnolia Ln    NY  southwest  8091-TTVAX
13         902      61     Highview Ter    NC  southeast  0280-XJGEX
14         903     878  Chamberlain Hwy    MA  southeast  5129-JLPIS
15         904     782    Kensington Rd    IA  southwest  3655-SNQYZ
16         905      34      Magnolia Ln    CA  northeast  8191-XWSZG

In [39]:
payment.head(5)

Payment_ID         Plan_ID             Payment_Method  Enrollment_Charges  \
0         301  60224GA0020005    Credit card (automatic)             1826.84   
1         302  33653ME0050002  Bank transfer (automatic)            11090.72   
2         303  29276OH0920418           Electronic check            39611.76   
3         304  20129IL0330020    Credit card (automatic)             1837.24   
4         305  37833WI0380028               Mailed check            10797.34   

  Average_Monthly_Tax_Credit  customerID  
0                      $265   8091-TTVAX  
1                      $237   0280-XJGEX  
2                      $259   5129-JLPIS  
3                      $307   3655-SNQYZ  
4                      $247   8191-XWSZG

In [40]:
tw_user.head(5)

Twitter_ID                  Name         Username  Verified  \
0   849775556        Health Options    CmtyHlthOptns     False   
1   467554336      Anthem BCBS News  AnthemBCBS_News     False   
2   573918122                Quartz               qz      True   
3    27922157                   AMA  AmerMedicalAssn      True   
4    20519602  FirstCareHealthPlans        firstcare     False   

                  Location                      url  No_of_Followers  \
0          Lewiston, Maine   http://t.co/PPlbfyzSyj              602   
1                            http://t.co/8eY95I0HOl             3372   
2                The World  https://t.co/gLfIrRbtbb           388622   
3  Chicago/Washington D.C.   http://t.co/MEg0g5QqL2           722458   
4                    Texas   http://t.co/jK5HXsqTnu              298   

   No_of_Tweets  Provider_ID  
0          3741          202  
1          9415          203  
2        176897          205  
3         27748          206  
4          1022          209

In [41]:
tweets_.head(5)

Tweet_ID  Twitter_ID  \
0  1253751279068864515   849775556   
1  1253385931744215043   849775556   
2  1233557662354300935   849775556   
3  1204443894957649927   849775556   
4  1169335101026709505   849775556   

                                                Text No_of Retweets  \
0  We are partnering with @PortlandChamber and ot...            NaN   
1          #leadershipskills https://t.co/6WD7eUprGl            NaN   
2  Are you worried about the spread of the corona...            NaN   
3  RT @MTUG_Maine: Please join us online! WED 12/...            NaN   
4  @JMHeffren Please contact Member Services at 8...            NaN   

   Favourites_Count  No_of_Retweets  
0                 0               0  
1                 0               0  
2                 0               0  
3                 0               2  
4                 0               0

# Use Cases

In [4]:
conn = MySQLdb.connect(host=HOST, port=3306, user=USER, passwd=PASSWD, db=DATABASE)
cursor=conn.cursor()

In [43]:
# USE CASE 1: Average age of all the customers 
cursor.execute("""SELECT AVG(Age) FROM customer;""")
rows = cursor.fetchall()
pprint.pprint(rows)

((Decimal('38.9108'),),)


In [44]:
# USE CASE 2: Name of the Customer who had the highest Enrollement Charges
cursor.execute("""select CustomerID from payment order by 'Enrollment Charges' desc limit 1;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('8091-TTVAX',),)


In [45]:
# USE CASE 3: What are people saying about insurance ? # Select all hashtags included in all tweets about insurance.
cursor.execute("""select text, created_at from twitter where text like "%insurance%" order by created_at;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('RT @CoverAlabama: Hundreds of thousands of Alabamians are counting on us '
  'for access to affordable health insurance. We MUST remain persiste…',
  datetime.datetime(2020, 4, 24, 22, 13, 41)),
 ('Florida Senator has 14 felony counts stemming from medicare insurance '
  'scams... yet old people on medicare vote for… https://t.co/7RTglvMrW7',
  datetime.datetime(2020, 4, 24, 22, 22, 27)),
 ('?? If you have transitioned to Medicare off of commercial insurance and can '
  'share tips, tricks, and watch-outs \n'
  '\n'
  '??… https://t.co/nioGrzkWXD',
  datetime.datetime(2020, 4, 24, 22, 32, 49)),
 ('@Shill_Destroyer @scrowder This whole "68,000 ppl die from no health '
  'insurance" line is pure bologna. Nobody NOBODY… https://t.co/SPjiQpS8X6',
  datetime.datetime(2020, 4, 24, 22, 33, 44)),
 ('@NshitThings @JoeBiden M4A is something Biden will never get.  He loves big '
  'insurance it is the reason he wants the… https://t.co/ranmTTqPMa',
  datetime.datetime(2020, 4, 24, 22, 42, 31)),

In [46]:
# USE CASE 4: Which State has the hieghtest number of Senior Citizens 

cursor.execute("""select state from customer_address ca inner join customer c on ca.customerID = c.customerID
where seniorcitizen !=  0 group by state order by count(*) desc limit 1;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('CA',),)


In [47]:
# USE CASE 5: Which is the most popular Insurance Provider 

cursor.execute("""select Issuer_Name from insuranceprovider ip inner join policy p on ip.provider_ID = p.provider_ID
group by Issuer_Name order by count(*) desc limit 1;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('Medica',),)


In [48]:
#USE CASE 6: Number of Female Customers who have medicaid enrollement

cursor.execute("""select count(*) as Number_of_Female_Medcaid_Benifiters from policy p inner join customer c
on p.plan_id = c.plan_id where c.sex like 'female' and Medicaid_Enrollment != 0;""")
rows = cursor.fetchall()
pprint.pprint(rows)

((81,),)


In [50]:
# USE CASE 7: Customer details who recieves the highest Yearly Tax Credit 

cursor.execute("""select c.customerID, c.age, c.sex, ca.number, ca.street, ca.state, ca.region from customer c
inner join customer_address ca on c.customerID = ca.customerID inner join payment p on c.plan_id = p.plan_id
order by (p.average_monthly_tax_credit * 12) desc limit 1;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('8091-TTVAX', 23, 'male', 26, 'Magnolia Ln', 'NY', 'southwest'),)


In [51]:
# USE CASE 8: Average Monthly Tax credit for  ****** Policy 

cursor.execute("""select average_monthly_tax_credit from payment p inner join policy po on p.plan_id = po.plan_id
where po.plan_marketing_name like '%Oscar Classic Silver%';;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('$294',),)


In [52]:
# USE CASE 9: Which is the most popular insurance policy among the southwest region 

cursor.execute("""select po.plan_marketing_name from policy po inner join customer c on po.plan_id = c.plan_id
inner join customer_address ca on c.customerID = ca.customerID where ca.region like 'southwest' group by c.plan_id
order by count(*) desc limit 1;;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('CareSource Marketplace Low Deductible Silver Dental, Vision, & Fitness',),)


In [53]:
# USE CASE 10: Which Insurance provider holds the heighest number of claims 
cursor.execute("""select Issuer_Name from insuranceprovider ip inner join policy p on ip.provider_ID = p.provider_ID
inner join claims c on p.plan_id = c.plan_id group by Issuer_Name order by count(*) desc limit 1;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('Medica',),)


In [54]:
# USE CASE 11: Number of top 10 latest tweets with date
cursor.execute("""SELECT created_at, COUNT(tweet_id) FROM twitter GROUP BY created_at ORDER BY COUNT(tweet_id) DESC LIMIT 10;""")
rows = cursor.fetchall()
pprint.pprint(rows)

((datetime.datetime(2020, 4, 24, 22, 52, 59), 3),
 (datetime.datetime(2020, 4, 24, 22, 15, 17), 3),
 (datetime.datetime(2020, 4, 24, 22, 41, 20), 3),
 (datetime.datetime(2020, 4, 24, 22, 25, 6), 3),
 (datetime.datetime(2020, 4, 24, 22, 31, 34), 3),
 (datetime.datetime(2020, 4, 24, 22, 53, 37), 2),
 (datetime.datetime(2020, 4, 24, 22, 55, 25), 2),
 (datetime.datetime(2020, 4, 24, 22, 56, 47), 2),
 (datetime.datetime(2020, 4, 24, 23, 0, 48), 2),
 (datetime.datetime(2020, 4, 24, 22, 36, 28), 2))


In [57]:
# USE CASE 12: What social media users are like other social media users in your domain?  #here we look for the Tweet with highest number of retweets
cursor.execute("""select t.text, t.tweet_id, t.no_of_retweets from tweets t order by no_of_retweets desc limit 10;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(("RT @nationaltrust: We're aiming to open many of our gardens and parks for "
  'free during this difficult time, so the nation can use open space…',
  1240056616264892419,
  14161),
 ('RT @LouisianaGov: Part of our new normal will mean wearing a mask or '
  'face-covering in public. Wearing a mask in public is just like holding…',
  1253359742052970498,
  509),
 ('Do not ingest or #InjectDisinfectant products – it will cause bodily harm. '
  'Treatments for #COVID19 must be guided b… https://t.co/YdjQmNZ70V',
  1253756569831047169,
  200),
 ('RT @DEAHQ: .@DEAHQ launches the #SecureYourMeds campaign, asking Americans '
  'to keep prescription medications safe and secure until they can…',
  1253358607107842048,
  85),
 ('RT @OHAOregon: There are currently no medications or treatments that are '
  'effective at preventing or treating #COVID19 and many of the home…',
  1253718059761598465,
  80),
 ('RT @BishopStAlbans: A message from Bishop Alan to the people of the Diocese '
  'of St Al

In [58]:
## USE CASE 13: What people, places or things are popular in your domain? # We can check the no. of retweets count of each text to determine the popular people in our domain

cursor.execute("""select pt.Name, t.no_of_retweets from tweets t inner join provider_twitter pt on t.twitter_id = pt.twitter_id
order by no_of_retweets desc limit 10;""")
rows = cursor.fetchall()
pprint.pprint(rows)


(('Lyndsey de Mestre QC', 14161),
 ('Louisiana Blue Cross', 509),
 ('AMA', 200),
 ('Louisiana Blue Cross', 85),
 ('PacificSource', 80),
 ('Lyndsey de Mestre QC', 46),
 ('BCBSSC', 32),
 ('BCBSSC', 27),
 ('Lyndsey de Mestre QC', 26),
 ('BCBSSC', 26))


In [59]:
## USE CASE 14: Who is the most influencer to posts? 
cursor.execute("""select pt.Name, t.Favourites_Count from provider_twitter pt inner join tweets t 
on pt.twitter_id = t.twitter_id where t.no_of_retweets !=0 and t.Favourites_Count !=0 group by t.text order by t.Favourites_Count DESC limit 10;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('AMA', 380),
 ('Kaiser Permanente', 39),
 ('Kaiser Permanente', 33),
 ('Kaiser Permanente', 25),
 ('AMA', 20),
 ('AMA', 17),
 ('AMA', 17),
 ('Arkansas Blue Cross and Blue Shield', 17),
 ('Kaiser Permanente', 14),
 ('CareSource', 12))


In [60]:
## USE CASE 15 : Number of tweets about each company 
cursor.execute("""SELECT p.issuer_name, COUNT(pt.no_of_tweets) as 'Number of Tweets' FROM insuranceprovider p
JOIN provider_twitter pt ON p.provider_id = pt.provider_id GROUP BY p.issuer_name ORDER BY COUNT(pt.no_of_tweets) DESC
LIMIT 10;""")
rows = cursor.fetchall()
pprint.pprint(rows)


(('Community Health Options', 1),
 ('Anthem Blue Cross and Blue Shield', 1),
 ('Quartz', 1),
 ('Medica', 1),
 ('FirstCare Health Plans', 1),
 ('Molina Healthcare', 1),
 ('Blue Cross and Blue Shield of Texas', 1),
 ('CareSource', 1),
 ('Ambetter from MHS', 1),
 ('Bright Health', 1))


In [63]:
## USE CASE 16: Which customer has the hieghest Claim amount
cursor.execute("""select c.customerid, c.age, ca.state, ca.region, cl.total_claim_amount from customer c 
inner join customer_address ca on c.customerid = ca.customerid inner join claims cl on c.plan_id = cl.plan_id
order by cl.total_claim_amount desc limit 5;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('2080-SRCDE', 30, 'MS', 'northeast', 154740),
 ('6865-JZNKO', 59, 'IN', 'southeast', 133800),
 ('7233-PAHHL', 59, 'MN', 'northeast', 122500),
 ('6067-NGCEU', 35, 'PA', 'northwest', 116900),
 ('9167-APMXZ', 48, 'VA', 'southeast', 116900))


In [64]:
##  USE CASE 17: Which Insurance Has the highest number of reports missing 
cursor.execute("""select p.issuer_name from insuranceprovider p inner join policy po on p.provider_id = po.provider_id
inner join claims c on po.plan_id = c.plan_id and c.report_available like 'YES' group by p.issuer_name order by count(*) desc
limit 5;""")
rows = cursor.fetchall()
pprint.pprint(rows)

(('Medica',),
 ('Anthem Blue Cross and Blue Shield',),
 ('Ambetter from MHS',),
 ('Blue Cross and Blue Shield of Texas',),
 ('Blue Cross Blue Shield Healthcare Plan of Georgia',))


In [ ]:
##  USE CASE 18 :What social media users are like other social media users in your domain?

cursor.execute("""select twitter_id, name, username, location, no_of_followers, verified from provider_twitter;""")
rows = cursor.fetchall()
pprint.pprint(rows)

## Optimizing the Database:

Database quering can be costly affair particulary when we are dealing with high number of joins and filter conditions. The more
the number tables involved in producing the result the more the time is needed to query all the matching data. This can be impacting the performance and there are several ways to optimise a database.

## Partitioning:

Patitioning is the process of splitting tables by physically putting tables on individual disk drives.Doing so will allow the database to subdivided into smaller pieces thus queries shall access only a fraction of the data can run faster because there is fewer data to scan.

In [5]:
# Applying Partition to table
cursor.execute("""ALTER TABLE insuranceprovider PARTITION BY HASH(provider_id) PARTITIONS 6;""")
rows = cursor.fetchall()
pprint.pprint(rows)

()


In [9]:
# Applying Partition to customer table
cursor.execute("""ALTER TABLE customer PARTITION BY HASH(age) PARTITIONS 6;""")
rows = cursor.fetchall()
pprint.pprint(rows)

()


In [11]:
# Applying Partition to customer_address table
cursor.execute("""ALTER TABLE customer_address PARTITION BY HASH(address_id) PARTITIONS 6;""")
rows = cursor.fetchall()
pprint.pprint(rows)

()


In [12]:
# Applying Partition to payment table
cursor.execute("""ALTER TABLE payment PARTITION BY HASH(payment_id) PARTITIONS 6;""")
rows = cursor.fetchall()
pprint.pprint(rows)

()


In [20]:
cursor.execute("""ALTER TABLE claims PARTITION BY HASH(`Claim-ID`) PARTITIONS 6;""")
rows = cursor.fetchall()
pprint.pprint(rows)

()


## Indexing:

The most effective and simplest way is to use the indexing.Indexes are columns that are applied on the table structure to speed up any action on the table, like select, update etc in the table. An best example is like that of an index in a book. Users wont be able to see the index but can see the perfomance imporved post applying the indexing.

To make indexing even more optimised, only create indexes on columns that will be frequently searched against.

In [ ]:
##  Applying Index to Tables:
cursor.execute("""create index idx_provider on insuranceprovider (provider_id);""")
rows = cursor.fetchall()
pprint.pprint(rows)

<img src= "Capture.png"/>

Since applying the indexing on the Insurance Provider table, executing of any queries that involving the indexed column shall be executed faster than before. Hence any operation that involves the indexed column like select, update insert shall all be better optimized.

## AUDIT:

AUDIT VALIDITY/ACCURACY:
	We say data is accurate only when it is neat and with no null or junk values. Here we used dropna() function to drop all the if they data holds any null values.
    
AUDIT COMPLETNESS:
  The database for an organization wherein we might be adding additional tables and columns depending on the real-world scenario. For now the database has been built in-par with the conceptual model.
    
AUDIT CONSISTENCY/UNIFORMITY:
	The datasets which have been used in this assignment show a uniform relationship between each of the dataset since they are linked to each other by a common attribute.


# CONTIBUTORS :

#### ASHWIN JOHN CHEMPOLIL, CRISPIN SUJITH CLETUS, VIGNESH THANIGAI SIVABALAN

## CITATIONS:

Sources from where you have gained knowledge or used codes, data. It may include Web links, github links, code taken from somewhere etc.

https://stackoverflow.com/questions/16476413/how-to-insert-pandas-dataframe-via-mysqlengine-into-database
https://docs.sqlalchemy.org/en/13/core/engines.html
https://stackoverflow.com/questions/43942357/pandas-to-sql-inserting-index
https://pythondata.com/collecting-storing-tweets-python-mysql/
https://github.com/nikbearbrown/INFO_6210/blob/master/Project/Jobs_DB_Project/IPYNB/MySQL.ipynb
https://github.com/nikbearbrown/INFO_6210/blob/master/Project/Jobs_DB_Project/IPYNB/NoSQL-Project-Social-Media.ipynb
https://www.alexkras.com/how-to-get-user-feed-with-twitter-api-and-python/
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html
https://towardsdatascience.com/twitter-data-collection-tutorial-using-python-3267d7cfa93e
https://stackoverflow.com/questions/34682828/extracting-specific-selected-columns-to-new-dataframe-as-a-copy


<copyright 2019 mavi>

# LICENSE

MIT License

Copyright (c) 2020 ASHWIN JOHN CHEMPOLIL, CRISPIN SUJITH CLETUS, VIGNESH THANIGAI SIVABALAN

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.